packages/webview_flutter/webview_flutter_web/lib/src/web_webview_controller.dart

```diff
@@ -106,6 +111,31 @@ class WebWebViewController extends PlatformWebViewController {
       encoding: encoding,
     ).toString();
   }
+
+  @override
+  Future<void> addJavaScriptChannel(
+    JavaScriptChannelParams javaScriptChannelParams,
+  ) async {
+    void handler(html.Event event) {
+      if (event is html.MessageEvent) {
+        javaScriptChannelParams.onMessageReceived(
+            JavaScriptMessage(message: event.data.toString()));
+      }
+    }
+
+    javascriptChannels[javaScriptChannelParams.name] = handler;
+    html.window.addEventListener('message', handler);
```

Comment:
If a programmer registers multiple 'message' handlers (one per `name`), this means that every `onMessageReceived` registered will be called for every message received, right? I think this is potentially expensive. Also calling `.toString()` on the object that is being passed around is not great, you lose the ability to move around [Transferable objects](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Transferable_objects#supported_objects), which might be interesting to users of this API.

Reply:
> If a programmer registers multiple 'message' handlers (one per name), this means that every onMessageReceived registered will be called for every message received, right? I think this is potentially expensive. 

That is true. This can easily happen when you have multiple webviews in a single application. Since the webviews do not know anything from each other, we did not see another option. I think if you have a low amount of handlers (e.g. a handful), there is no performance hit.

> Also calling .toString() on the object that is being passed around is not great, you lose the ability to move around Transferable objects, which might be interesting to users of this API.

On Android and iOS it is only possible to send string messages. We tried to mimic that API by limiting the  allowed message types to string-values only. This way everything can be handled the same way in Flutter (i.e. all incoming messages are always string). If one wants to send complex messages one can always send a stringified object.